### ISSUE TITLE
[cml][sdw][STRA-P]:  The volume of built-in audio output is low as come back from runtime-pm

### ISSUE THREAD

**Issue by jason77-wang** [thesofproject/linux#2605]
On the STRA-P (cml soundwire audio), before runtime suspend, the audio (speaker/headphone)'s output volume is correct, after it enters runtime suspend and runtime resume, the output volume is lower than before for about 1-2 seconds,then it changes back to normal.

This issue could be reproduced with sof-firmware-v1.6 and the kernel branch sof-dev-rebase. This issue could not be reproduced on MLK OLY machine (TGL soundwire audio).

**Comment by keyonjie** [comment]
@singalsu is this expected for volume ramping?

**Comment by singalsu** [comment]
@jason77-wang A recording of this issue would be really good to have. Are you able to make such? Acoustical recording is also OK.

**Comment by singalsu** [comment]
@keyonjie None of the defined ramps is 1-2s long, though technically it would be possible with wrong made topology.

**Comment by plbossart** [comment]
@jason77-wang are you using the same userspace components on both devices? There's really nothing specific to CML or TGL here.

**Comment by jason77-wang** [comment]
@plbossart  yes, same userspace components on both devices.

@singalsu  OK, will record a sound sample, since I don't have this machine, need my colleague in Taipei office to do that.

Thanks.

**Comment by lihow731** [comment]
I put the record video in blew link:
[The first sound effect is lower volume](https://drive.google.com/file/d/1AH0ddFXrixzmBpuIK4auyfaY6qPg3sJ3/view?usp=sharing)

**Comment by singalsu** [comment]
Yes the issue can be heard and seen (below) but please make a recording with a steady sine wave. The key presses are too much away to see the gain envelope. This recording quality is totally sufficient, just different signal.

![Screenshot from 2020-12-08 11-17-25](https://user-images.githubusercontent.com/28754793/101464443-51a28a00-3947-11eb-88c6-de3ae1709a09.png)

[sine.zip](https://github.com/thesofproject/linux/files/5658164/sine.zip)

**Comment by keyonjie** [comment]
@aiChaoSONG @WYL-12 can you help to reproduce this?

**Comment by aiChaoSONG** [comment]
@jason77-wang @lihow731 I test this issue with sof-firmware-v1.6 and the sof kernel sof-dev-rebase-20201201 with `speaker-test -Dhw:0,2 -t wav -c 2`, I can hear that the `fron` in front-left is a little bit low, but from the t sound in front, the volume go back to normal. after checking the waveform of front-left.wav, I find that the `fron` duration is about 0.2s, so I think this is a normal fade-in. could you please play a continuous sound or music to see if it is really a 1~2s low volume period and also confirm my test result? I mean the system `ding` may be to short for the verification.

**Comment by keyonjie** [comment]
@jason77-wang @lihow731 can you confirm the volume ramping time is actually only around 0.2s? if so, this is expected.

**Comment by singalsu** [comment]
The token SOF_TKN_VOLUME_RAMP_STEP_MS is set to 20 for 20 ms long ram in playback topologies. If the ramp is 0.2s it's too long. The continues sine wave recording is needed to check the ramp characteristic.

**Comment by singalsu** [comment]
If you don't have a sine wave clip, please use this:

[sine_997hz_6db_10s_mono_16b.wav.gz](https://github.com/thesofproject/linux/files/5723887/sine_997hz_6db_10s_mono_16b.wav.gz)

**Comment by jason77-wang** [comment]
Since I don't have the hw, I could not confirm if the ramp is 0.2s or 1~2s. Our QA filed a bug against this issue, that means the ramp on this machine is obvious longer than on other machines.

@lihow731, do you think it is 0.2s or 1~2s?

Thanks.

**Comment by lihow731** [comment]
Please watch the video that the sine-977Hz twice was played twice.

https://drive.google.com/file/d/18Q5SzKajcvzIVoXgemSyXxGuzOYRMZ3N/view?usp=sharing

**Comment by singalsu** [comment]
Thanks @lihow731 ! Here's the audio waveforms from your video. It was MPEG AAC format so not as reliable as uncompressed wav but it looks like the first ramp is 150 ms long and second 5 ms or less (AAC coding impacts, so I assume there's no ramp).

![Screenshot from 2020-12-22 10-21-20](https://user-images.githubusercontent.com/28754793/102866736-6c452a80-4440-11eb-9e4a-5ea71950da14.png)
![Screenshot from 2020-12-22 10-23-49](https://user-images.githubusercontent.com/28754793/102866755-749d6580-4440-11eb-88f2-03327673ec62.png)

@keyonjie @aiChaoSONG Do  you know which topology this device uses? The playback start ramp should be 20 ms in SOF topologies. The 150 ms seen is too long.

@lihow731 If you can check it, the used topology name can be found from console with `dmesg | grep "loading topology"` .

**Comment by aiChaoSONG** [comment]
[stra_loopback.zip](https://github.com/thesofproject/linux/files/5728834/stra_loopback.zip)

**Comment by keyonjie** [comment]
it could be the ramp time for the codec side?

**Comment by singalsu** [comment]
@aiChaoSONG The playback start ramp should be 20 ms. I wonder if the loopback was made with pulseaudio since there wasn't really any start ramp, mainly a DAC impulse response for sine wave start.

When looking at sof-cml-rt711-rt1308-rt715.conf there's some variation in playback pga's SOF_TKN_VOLUME_RAMP_STEP_MS values. I'll check from .conf that they get the correct values.

@keyonjie Yep, the codec could do this.

**Comment by singalsu** [comment]
There's a mistake in muxdemux pipeline volume tokens. Need to use the macros with pipeline IDs to make tokens unique to that volume instance. I wonder if the 250 comes from capture side where it's a normal value.

![sof-cml-rt711-rt1308-rt715](https://user-images.githubusercontent.com/28754793/102877481-7d496800-444f-11eb-9dc3-e5deb55c8875.png)

PGA1.0 is 20ms
PGA3.1 is 250ms
PGA6.0 is 20ms
PGA7.0 is 20ms
PGA8.0 is 20ms

I can do a PR to fix this today but I can't test it (don't have the device, will start my holidays in few hours and back in Jan 4th). I can attach a binary topology that could be tried on the device by copying it over over original. 

(The bug severity in SOF topology is not critical in my opinion.)
